numpy/lib/arraysetops.py

```diff
@@ -239,59 +300,119 @@ def reshape_uniq(uniq):
         uniq = np.swapaxes(uniq, 0, axis)
         return uniq
 
-    output = _unique1d(consolidated, return_index,
-                       return_inverse, return_counts)
-    if not (return_index or return_inverse or return_counts):
-        return reshape_uniq(output)
+    consolidated_base = np.array_view_root(consolidated)
+    orig_base = np.array_view_root(orig_ar)
+    if return_index or return_inverse:
+        # We won't perform an in-place sort even if requested
+        mask_is_sorted = assume_sorted
+    elif orig_base is consolidated_base:
```

Comment:
This part of the change makes me uneasy - can we leave out this `.base` traversal to avoid a copy to a different PR? It would be nice to get `return_mask` into `1.14.

Reply:
This part is key though, without it, copies still have to be made, and that increases memory usage many-fold

Reply:
Would it be ok to make it opt-in? (say, `try_inplace = True`)

Reply:
I'd still rather separate the new feature (`return_mask`) and the `.base`-checking optimization into separate PRs - I'd want a second opinion on the latter half, which will be easier to get if the diff is smaller